### PR TITLE
Discard fast methods coming from trace_methods

### PR DESCRIPTION
**Pull request by eyalkoren**
Closes elastic/apm-agent-java#440 

Made a few decisions:
1. Never discard transactions
2. Never discard branches leading to async branches so not to break span parent referencing
3. Off by default (config option set to 0)
4. When on (ie threshold > 0) - branches leading to spans ending with Exceptions are not discarded

The last one is explicitly related to a threshold setting, however, it makes sense that if someone opts to lose non-interesting, error traces are still valuable

### CODE REVIEW

File: apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/CoreConfiguration.java [hunk 1]
```diff
@@ -295,6 +292,25 @@
...
+            "duration. When set to a value greater than 0, spans representing methods traced based on `trace_methods` will be discarded " +
+            "by default.\n" +
+            "Such methods will be traced and reported if one of the following applies:\n" +
+            " - This method's duration crossed the configured threshold.\n" +
+            " - This method ended with Exception.\n" +
+            " - A method executed as part of the execution of this method crossed the threshold or ended with Exception.\n" +
+            " - A \"forcibly-traced method\" (e.g. DB queries, HTTP exits, custom) was executed during the execution of this method.\n" +
+            "Set to 0 to disable.\n" +
+            "\n" +
+            "NOTE: Transaction are never discarded, regardless of their duration. This configuration affects only spans.\n" +
```
Comment (felixbarny) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/CoreConfiguration.java]:
```suggestion
            "NOTE: Transactions are never discarded, regardless of their duration. This configuration affects only spans.\n" +
```

File: apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/CoreConfiguration.java [hunk 2]
```diff
@@ -295,6 +292,25 @@
...
+            "In order not to break span references, all spans leading to an async operations are never discarded, regardless \n" +
+            "of their duration.\n")
+        .buildWithDefault(TimeDuration.of("0ms"));
```
Comment (felixbarny) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/CoreConfiguration.java]: Should we enhance TimeDuration to add support for `us` (µs)?

File: apm-agent-plugins/apm-java-concurrent-plugin/src/test/java/co/elastic/apm/agent/concurrent/AsyncTraceMethodInstrumentationTest.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        new TestAsyncTraceMethodsClass().invokeAsync();
+        assertThat(reporter.getTransactions()).hasSize(1);
+        assertThat(reporter.getSpans()).hasSize(4);
+    }
+
+
+    private static class TestAsyncTraceMethodsClass {
+
+        /**
+         * Calling this method results in this method call tree:
```
Comment (felixbarny) [inline apm-agent-plugins/apm-java-concurrent-plugin/src/test/java/co/elastic/apm/agent/concurrent/AsyncTraceMethodInstrumentationTest.java]: seems like the formatting got messed up, maybe your local history still contains the old version

File: apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/CoreConfiguration.java (hunk 2 above)
Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/CoreConfiguration.java]: Not sure if sub-millisecond is enhancing enough. I imagine someone using `trace_methods` with wildcards and hoping to get performance costs down shouldn't configure microseconds thresholds.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/CoreConfiguration.java [hunk 4]
```diff
@@ -261,7 +258,7 @@
     private final ConfigurationOption<List<MethodMatcher>> traceMethods = ConfigurationOption
         .builder(new ListValueConverter<>(MethodMatcherValueConverter.INSTANCE), List.class)
         .key("trace_methods")
-        .tags("added[1.3.0,Enhancements in 1.4.0]")
+        .tags("added[1.3.0,Enhancements in 1.4.0 and 1.6.0]")
```
Comment (felixbarny) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/CoreConfiguration.java]: `1.7.0` now
But I'm not sure if that can be considered an enhancement to the `trace_methods` directly.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java [hunk 5]
```diff
@@ -293,6 +296,14 @@ void setRecorded(boolean recorded) {
         }
     }
 
+    public void setDiscard(boolean discard) {
+        this.discard = discard;
```
Comment (felixbarny) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java]: We have a similar flag in `Transaction` called `noop`. Maybe we should rather pull that up to `AbstractSpan`?

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ElasticApmTracer.java [hunk 6]
```diff
@@ -475,6 +475,10 @@ public void deactivate(TraceContextHolder<?> holder) {
         }
         final Deque<TraceContextHolder<?>> stack = activeStack.get();
         assertIsActive(holder, stack.poll());
+        if (!stack.isEmpty() && !holder.isDiscard()) {
+            //noinspection ConstantConditions
+            stack.peek().setDiscard(false);
```
Comment (felixbarny) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ElasticApmTracer.java]: don't we have to set discard to false on all elements in the stack?

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java [hunk 7]
```diff
@@ -95,6 +95,8 @@ public boolean asChildOf(TraceContext child, Object ignore) {
     private final Id transactionId = Id.new64BitId();
     private final StringBuilder outgoingHeader = new StringBuilder(TRACE_PARENT_LENGTH);
     private byte flags;
+    private boolean discard;
```
Comment (felixbarny) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java]: also copy on `asChildOf`?

**Comment by eyalkoren** [comment]
> I think this approach may not be generic enough as there might be async spans without a prior wrapping.

Isn't this the same with reference counting, which relies on wrapping as well?

> Also, async spans which don't outlive their parent can safely be discarded when they are fast.

May not be consistent - can sometimes outlive parent and sometimes not.

In general, this PR is an enhancement for `trace_methods` users that use it with wildcards. It is still not the recommended way of using this feature.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/CoreConfiguration.java (hunk 4 above)
Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/CoreConfiguration.java]: > But I'm not sure if that can be considered an enhancement to the `trace_methods` directly.

Why is that? It is only related to `trace_methods` and has no effect when not used. It extends the configuration capabilities of this feature.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java (hunk 5 above)
Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java]: Seems like currently they effectively serve a similar purpose. I will move it and if we see a new need to differentiate `noop` from `discard` we can reiterate later.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ElasticApmTracer.java (hunk 6 above)
Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ElasticApmTracer.java]: Not really. All spans are created as non-discardable, except from the `trace_methods` spans. This means there is no race between `deactivate` and `end` and it should be safe to do it this way.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java (hunk 7 above)
Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java]: No. Only `trace_methods` spans should be created as discardable. Their children shouldn't, unless the are such as well.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java (hunk 5 above)
Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java]: Actually, this breaks the logic used for this flag- `noop` transactions are not-sampled, and this is an inheritable trait, so all decedents will be non-sampled. However, marking a span as `discard` should not be inherited and should not affect parent (as opposed to the `discard` flag).

**Comment by eyalkoren** [comment]
Jenkins run the tests

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ElasticApmTracer.java (hunk 6 above)
Comment (felixbarny) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ElasticApmTracer.java]: gotcha

**Pull request merged**